lib/AnalysisStructured/PtrAnalysis.cpp

```diff
@@ -0,0 +1,1128 @@
+      if (auto attr = op.getValue().dyn_cast<DenseElementsAttr>()) {
+        if (!attr.isSplat()) {
+          InFlightDiagnostic diag = emitError(loc)
+                                    << "other value used in masked load "
+                                       "produced by unsupported instruction";
+          return nullptr;
+        }
+        auto elemValue = attr.getSplatValue<Attribute>();
+        auto constOp = arith::ConstantOp::materialize(
+            builder, elemValue, attr.getElementType(), op.getLoc());
+        return reconstructScalarValue(constOp.getResult());
+      }
+    } else if (auto op = operand.getDefiningOp<triton::SplatOp>()) {
+      operand = op.getSrc();
+    } else if (auto op = operand.getDefiningOp<arith::SIToFPOp>()) {
+      ops.push_back(op.getOperation());
+      operand = op.getIn();
+    } else if (auto op = operand.getDefiningOp<arith::TruncFOp>()) {
+      ops.push_back(op.getOperation());
+      operand = op.getIn();
+    } else {
+      InFlightDiagnostic diag = emitError(loc)
+                                << "other value used in masked load produced "
+                                   "by unsupported instruction";
+      return nullptr;
+    }
+  }
+  return nullptr;
+}
+
+namespace triton {
+
+int32_t PtrSState::getRank() const {
+  assert(offsets.size() == sizes.size() && offsets.size() == strides.size() &&
+         modulos.size() == offsets.size());
+  return offsets.size();
+}
+
+bool PtrSState::isEmpty() const {
+  return (getRank() == 0 && !source && !scalar);
+}
+
+bool PtrSState::hasModulo() const {
+  for (int32_t i = 0; i < getRank(); i++) {
+    if (dimHasModulo(i)) {
+      return true;
+    }
+  }
+  return false;
+}
+
+bool PtrSState::dimHasModulo(uint32_t dim) const {
+  assert(dim < getRank());
+
+  auto intAttr = getIntAttr(modulos[dim]);
+  if (!intAttr.has_value()) {
+    return true;
+  }
+
+  return intAttr.value() != 0;
+}
+
+LogicalResult PtrSState::addState(const PtrSState &lhsState,
+                                 const PtrSState &rhsState, Operation *op,
+                                 OpBuilder &builder) {
+  assert(isEmpty() && lhsState.getRank() == rhsState.getRank());
+  auto loc = op->getLoc();
+
+  if (lhsState.source && rhsState.source) {
+    op->emitRemark(
+        "PtrAnalysis: do not support adding two pointer states that both "
+        "have base pointers");
+    return failure();
+  }
+
+  source = lhsState.source ? lhsState.source : rhsState.source;
+
+  // AddPtr where both lhs and rhs containing modulo operators not supported
+  if (lhsState.hasModulo() && rhsState.hasModulo()) {
+    op->emitRemark("PtrAnalysis: do not support adding two pointer states "
+                   "that both have modulo");
+    return failure();
+  }
+
+  if (lhsState.hasModulo() || rhsState.hasModulo()) {
+    // visitOperandSplat and visitOperandExpandDims should enforce below
+    assert(lhsState.getRank() <= 2);
```

Comment:
I realized that this is to guard the reasoning that `visitOperandSplat and visitOperandExpandDims should enforce` that this rank is true, which they do afaik. Leaving it in since hitting this condition actually indicates that something bad has happened.